Add Full-duplex serial driver for ARM boards (#9842)
qmk/qmk_firmware: docs/serial_driver.md

                                   //  1: about 230400 baud (default)
                                   //  2: about 115200 baud
                                   //  3: about 57600 baud
                                   //  4: about 38400 baud
                                   //  5: about 19200 baud
#define SERIAL_USART_DRIVER UARTD1 // USART driver of TX and RX pin. default: UARTD1
#define SERIAL_USART_TX_PAL_MODE 7 // Pin "alternate function", see the respective datasheet for the appropriate values for your MCU. default: 7
#define SERIAL_USART_RX_PAL_MODE 7 // Pin "alternate function", see the respective datasheet for the appropriate values for your MCU. default: 7
#define SERIAL_USART_TIMEOUT 100 // USART driver timeout. default 100
```

You must also enable the ChibiOS `UART` with blocking api feature:
* In your board's halconf.h: `#define HAL_USE_USAR TRUE` and `#define UART_USE_WAIT TRUE`
* In your board's mcuconf.h: `#define STM32_UART_USE_USARTn TRUE` (where 'n' matches the peripheral number of your selected USART on the MCU)

Do note that the configuration required is for the `UART` peripheral, not the `SERIAL` peripheral.

> #### Pins for USART Peripherals with Alternate Functions for selected STM32 MCUs

Reviewer: Would be nice if all of this was a single table, though I know what may be complex to do.

Author: Merged the tables into a single one for each mcu. Having one big table for all mcus is not practical imho.

USART2:
| Pin  | Function | Mode |
| ---- | -------- | ---- |
| PA2  | TX       | AF1  |
| PA3  | RX       | AF1  |
| PA14 | TX       | AF1  |
| PA15 | RX       | AF1  |

USART3:
| Pin  | Function | Mode |
| ---- | -------- | ---- |
| PB10 | TX       | AF4  |
| PB11 | RX       | AF4  |
| PC4  | TX       | AF1  |
| PC5  | RX       | AF1  |
| PC10 | TX       | AF1  |
| PC11 | RX       | AF1  |
| PD8  | TX       | AF0  |
| PD9  | RX       | AF0  |

USART4:
| Pin | Function | Mode |
| --- | -------- | ---- |
| PA0 | TX       | AF4  |
| PA1 | RX       | AF4  |

##### STM32F103 Medium Density (C8-CB) [Datasheet](https://www.st.com/resource/en/datasheet/stm32f103c8.pdf)

Pin Swap available: N/A

TX Pin is always Alternate Function Push-Pull, RX Pin is always regular input pin for any USART peripheral. **For STM32F103 no additional Alternate Function configuration is necessary. QMK is already configured.**

Pin remapping:

The pins of USART Peripherals use default Pins that can be remapped to use other pins using the AFIO registers. Default pins are marked **bold**. Here is an example how you can remap the pins for USART1.

```c
// Remap USART1 from pins PA9/PA10 to PB6/PB7

> void keyboard_pre_init_user(void) {

Reviewer: Would rather this be a setting, via a define, than this.

Author: Add defines for USART remapping. This affects half-duplex as well as full-duplex.

* In your board's mcuconf.h: `#define STM32_SERIAL_USE_USARTn TRUE` (where 'n' matches the peripheral number of your selected USART on the MCU)

Do note that the configuration required is for the `SERIAL` peripheral, not the `UART` peripheral.

### USART Full-duplex

> Targeting STM32 boards where communication is offloaded to a USART hardware device. The advantage over bitbang is that this provides fast and accurate timings. USART Full-Duplex requires two conductors instead of one conductor unlike the Half-duplex driver, but it is more efficent as it uses DMA transfers, which can result in even faster transmission speeds.

Author: No external pullup is needed. I could run with dual open drain pins, which requires two pull up resistors. That is not the default though.

> Targeting STM32 boards where communication is offloaded to a USART hardware device. The advantage over bitbang is that this provides fast and accurate timings. USART Full-Duplex requires two conductors instead of one conductor unlike the Half-duplex driver, but it is more efficent as it uses DMA transfers, which can result in even faster transmission speeds.

Author: Updated the wording.

> Targeting STM32 boards where communication is offloaded to a USART hardware device. The advantage over bitbang is that this provides fast and accurate timings. USART Full-Duplex requires two conductors instead of one conductor unlike the Half-duplex driver, but it is more efficent as it uses DMA transfers, which can result in even faster transmission speeds.

Reviewer: Awesome!

#### Pin configuration

> `SERIAL_USART_TX_PIN` is the USART `TX` pin, `SERIAL_USART_RX_PIN` is the USART `RX` pin. No external pull-up resistors are needed as the `TX` pin operates in push-pull mode. To use this driver the usart peripherals `TX` and `RX` pins must be configured with the correct Alternate-functions. If you are using an STM32F103 everything is already setup, for other MCUs which are using a more modern GPIO configuration you have to specify these by setting `SERIAL_USART_TX_PAL_MODE` and `SERIAL_USART_RX_PAL_MODE`. Refeer to the corresponding datasheets of your MCU or find those Settings in the table below.

Reviewer 2: Pretty sure we target Proton-C/F303 by default. Will need to confirm defaults elsewhere.
```suggestion
`SERIAL_USART_TX_PIN` is the USART `TX` pin, `SERIAL_USART_RX_PIN` is the USART `RX` pin. No external pull-up resistors are needed as the `TX` pin operates in push-pull mode. To use this driver the usart peripherals `TX` and `RX` pins must be configured with the correct Alternate-functions. If you are using a Proton-C, everything is already setup, for other MCUs which are using a more modern GPIO configuration you have to specify these by setting `SERIAL_USART_TX_PAL_MODE` and `SERIAL_USART_RX_PAL_MODE`. Refeer to the corresponding datasheets of your MCU or find those Settings in the table below.
```

Author: Added Proton-C to the sentence and adjusted the following, not mentioning the stm32f103 would have skewed the meaning as If the f303 wouldn't have a modern gpio configuration.

#### Connecting the halves and Pin Swap

> Please note that `TX` of the master halve has to be connected with the `RX` pin of the slave halve and `RX` of the master halve has to be connected with the `TX` pin of the slave halve! Usually this pin swap has to be done outside of the MCU e.g. with cables or on the pcb. Some MCUs like the STM32F303 allow this pin swap directly inside the MCU, this feature can be enabled using `#define SERIAL_USART_PIN_SWAP` in your config.h.

Reviewer 2: ```suggestion
Please note that `TX` of the master half has to be connected with the `RX` pin of the slave half and `RX` of the master half has to be connected with the `TX` pin of the slave half! Usually this pin swap has to be done outside of the MCU e.g. with cables or on the pcb. Some MCUs like the STM32F303 allow this pin swap directly inside the MCU, this feature can be enabled using `#define SERIAL_USART_PIN_SWAP` in your config.h.
```

#### Setup
To use the driver, add this to your rules.mk:

```make
SERIAL_DRIVER = usart_duplex
```

Next configure the hardware via your config.h:

```c
#define SERIAL_USART_TX_PIN B6     // USART TX pin
#define SERIAL_USART_RX_PIN B7     // USART RX pin
//#define USART1_REMAP             // Remap USART TX and RX pins on STM32F103 mcus, see table below.
//#define SERIAL_USART_PIN_SWAP    // Swap TX and RX pins if keyboard is master halve.
                                   // Check if this feature is necessary with your keyboard design and available on the mcu.

> #define SELECT_SOFT_SERIAL_SPEED 1 // or 0, 2, 3, 4, 5
>                                    //  0: about 460800 baud
>                                    //  1: about 230400 baud (default)
>                                    //  2: about 115200 baud
>                                    //  3: about 57600 baud
>                                    //  4: about 38400 baud
>                                    //  5: about 19200 baud

Reviewer 2: ```suggestion
#define SELECT_SOFT_SERIAL_SPEED 1 // or 0, 2, 3, 4, 5
                                   //  0: 460800 baud
                                   //  1: 230400 baud (default)
                                   //  2: 115200 baud
                                   //  3: 57600 baud
                                   //  4: 38400 baud
                                   //  5: 19200 baud
```

#define SERIAL_USART_DRIVER UARTD1 // USART driver of TX and RX pin. default: UARTD1
#define SERIAL_USART_TX_PAL_MODE 7 // Pin "alternate function", see the respective datasheet for the appropriate values for your MCU. default: 7
#define SERIAL_USART_RX_PAL_MODE 7 // Pin "alternate function", see the respective datasheet for the appropriate values for your MCU. default: 7
#define SERIAL_USART_TIMEOUT 100 // USART driver timeout. default 100
```

You must also enable the ChibiOS `UART` with blocking api feature:

> * In your board's halconf.h: `#define HAL_USE_USAR TRUE` and `#define UART_USE_WAIT TRUE`

Reviewer 2: ```suggestion
* In your board's halconf.h: `#define HAL_USE_UART TRUE` and `#define UART_USE_WAIT TRUE`
```

|    Pin     | Function | Mode |     USART_REMAP     |
| ---------- | -------- | ---- | ------------------- |
| **USART1** |          |      |                     |
| **PA9**    | TX       | AFPP |                     |
| **PA10**   | RX       | IN   |                     |
| PB6        | TX       | AFPP | USART1_REMAP        |
| PB7        | RX       | IN   | USART1_REMAP        |
| **USART2** |          |      |                     |
| **PA2**    | TX       | AFPP |                     |
| **PA3**    | RX       | IN   |                     |
| PD5        | TX       | AFPP | USART2_REMAP        |
| PD6        | RX       | IN   | USART2_REMAP        |
| **USART3** |          |      |                     |
| **PB10**   | TX       | AFPP |                     |
| **PB11**   | RX       | IN   |                     |
| PC10       | TX       | AFPP | USART3_PARTIALREMAP |
| PC11       | RX       | IN   | USART3_PARTIALREMAP |
| PD8        | TX       | AFPP | USART3_FULLREMAP    |
| PD9        | RX       | IN   | USART3_FULLREMAP    |

> ##### STM32F303 [Datasheet](https://www.st.com/resource/en/datasheet/stm32f303cc.pdf)

Reviewer 2: Move this to be first, given the Proton-C focus.
